src/t_stream.c

```diff
@@ -2301,11 +2301,18 @@ void xreadCommand(client *c) {
             }
             if (o) {
                 stream *s = o->ptr;
-                ids[id_idx] = s->last_id;
-                if (streamDecrID(&ids[id_idx]) != C_OK) {
-                    /* shouldn't happen */
-                    addReplyError(c,"the stream last element ID is 0-0");
-                    goto cleanup;
+                if (s->length) {
+                    /* We need to get the last valid ID.
+                     * It is impossible to use s->last_id because
+                     * entry with s->last_id may have been removed. */
+                    streamLastValidID(s, ids+id_idx);
+                    if (streamDecrID(&ids[id_idx]) != C_OK) {
+                        /* shouldn't happen */
+                        addReplyError(c,"the stream last element ID is 0-0");
+                        goto cleanup;
+                    }
+                } else {
+                    ids[id_idx] = s->last_id;
```

Comment:
is `if (streamDecrID(&ids[id_idx]) != C_OK)` needed here?

Reply:
If you mean `if` statement then i think it is not needed, because there are no cases when `last_id` is `0-0` (at least i can't find any). But it have been added in initial commit of this feature, so I've decided to leave it as it was.
But we definitely need `streamDecrID` call, because this id will be incremented later.

Reply:
you can use `XGROUP CREATE mystream mygroup $ MKSTREAM` to create a stream whose last_id is 0-0.

Reply:
hm, so there was another bug. In this PR it is gone, because now this call is protected with `if (s->length)` and for non-empty stream `last_id` cannot be `0-0`.
I've added test for this case. It fails on `unstable` but passes on this PR.

Reply:
@edkalina maybe you're right, ping @oranagra

Reply:
instead of this whole block, why not change
```
ids[id_idx] = s->last_id;
```
to
```
streamLastValidID(s, &ids[id_idx]);
```

Reply:
ah, sorry, get it wrong. I was talking about `streamDecrID()`. Yes, tests are not failing with this simple change.
I've wrapped `streamLastValidID()` call with `if (s->length)` for two reasons:
1. I didn't investigate behaviour of `streamLastValidID`, so I don't know the result of its call for empty stream
2. I can see that each call of `streamLastValidID()` is wrapped with `if (s->length)` in `t_stream.c`
My guess was that `ids[id_idx]` will not be set in case of empty stream and will contain garbage, so checking for stream's length was added.

Reply:
ok, I did more investigations and now it is clear for me. My assumption was right, id will contain garbage from allocated memory. And it is ok for almost all cases:
* if this read is not blocking then `nil` will be returned, that is correct value
* if this read is blocking then `xread` will be re-processed with the same args when new value will be added. Then stream will not be empty and `xread` will return correct value.

The only bad case is if id gets zeros from uninitialized memory and command will fail on `if (streamDecrID(&ids[id_idx]) != C_OK)`. But I think this check can be safely removed:
```c
streamLastValidID(s, ids+id_idx);
streamDecrID(&ids[id_idx]);
```

Tests are ok with this patch.
I've also checked case with zeros (added preinitialization of ids with zeros). Some tests fail with `if` and all pass without `if`.

wdyt? did I miss smth?

Reply:
i feel a bit uncomfortable leaving `ids[id_idx]` uninitialized (even though it's not referenced later on)
let's go with:
```
if (o && ((stream *)o->ptr)->length) {
    stream *s = o->ptr;
    streamLastValidID(s, &ids[id_idx]);
    streamDecrID(&ids[id_idx]);
} else {
    ids[id_idx].ms = 0;
    ids[id_idx].seq = 0;
}
```